src/fancy-transformers/defragmenter.ts

```diff
@@ -0,0 +1,87 @@
+import logging = require('../logging/logging');
+import arraybuffers = require('../arraybuffers/arraybuffers');
+import Fragment = require('./fragment');
+
+var log :logging.Log = new logging.Log('fancy-transformers');
+
+class Defragmenter {
+  private tracker_ : {[index: string]: ArrayBuffer[]}={};
+  private counter_ : {[index: string]: number}={};
+  private complete_: string[]=[];
+
+  public constructor() {}
+
+  public addFragment(fragment:Fragment) {
+    var hexid=arraybuffers.arrayBufferToHexString(fragment.id);
+    if(hexid in this.tracker_) { // A fragment for an existing packet
+      var fragments : ArrayBuffer[]=this.tracker_[hexid];
+      if(fragments[fragment.index]!=null) { // Duplicate fragment
+        log.info('Duplicate fragment %1: %2 / %3', hexid, fragment.index, fragment.count);
+      } else { // New fragment
+        fragments[fragment.index]=fragment.payload;
+        this.tracker_[hexid]=fragments;
+        this.counter_[hexid]=this.counter_[hexid]-1;
+        if(this.counter_[hexid]==0) {
+          this.complete_.push(hexid);
+        }
+      }
+    } else { // A fragment for a new packet
+      var fragments : ArrayBuffer[]=[];
+      for(var i=0; i<fragment.count; i++) {
```

Comment:
I donno if you care about the difference but Array has a constructor that takes a single numerical arg, and results in an array of size `arg`, each initialized to `undefined`.